kernel/common/enumerator.rb

```diff
@@ -148,7 +149,19 @@ def rewind
     end
 
     def size
-      @size.kind_of?(Proc) ? @size.call : @size
+      if @size.kind_of?(Proc)
+        @size.call
+      elsif @size.is_a? Numeric
+        @size
+      elsif @size.respond_to? :size
+        @size.size
+      else
```

Comment:
@YorickPeterse in the case of each_with_index it was returning the array, there may be another (more correct) way to resolve this
```
[1, 2, 3, 5].each_with_index.size
-> [1, 2, 3, 5]
```